### PR TITLE
Increase DB/CPU perf of `_is_server_still_joined` check.

### PR DESCRIPTION
**Pull request by erikjohnston**
For rooms with large amount of state a single user leaving could cause us to go and load a lot of membership events and then pull out membership state in a large number of batches.

### CODE REVIEW

File: synapse/storage/persist_events.py [hunk 1]
```diff
@@ -735,9 +736,32 @@ def _maybe_start_persisting(self, room_id: str):
                 if event_id == event.event_id:
                     if event.membership == Membership.JOIN:
                         return True
+                else:
+                    events_to_check.append(event_id)
```
Comment (richvdh) [inline synapse/storage/persist_events.py]: this should surely only happen if the event is nowhere to be found in `ev_ctx_rm` (ie, if we hit the end of the `for` loop over `ev_ctx_rm`)

File: synapse/storage/persist_events.py [hunk 2]
```diff
@@ -735,9 +736,32 @@ def _maybe_start_persisting(self, room_id: str):
...
+        # Check if any of the changes that we don't have events for are joins.
+        if events_to_check:
+            rows = await self.main_store.get_membership_from_event_ids(events_to_check)
+            is_still_joined = any(row["membership"] == Membership.JOIN for row in rows)
+            if is_still_joined:
+                return True
+
+        # None of the changes to state are joins, so we fall back to checking
+        # the current state of the room to see if any of our users are joined
+        # (ingnoring users that have changed state.)
```
Comment (richvdh) [inline synapse/storage/persist_events.py]: "Ingnoring", but I would rephrase for clarity:

```suggestion
        # None of the new state events are joins, so now we check the current
        # room state to see if there are any other users in the room.
```

File: synapse/storage/data_stores/main/roommember.py [hunk 3]
```diff
@@ -868,6 +868,37 @@ def get_membership_from_event_ids(
             desc="get_membership_from_event_ids",
         )
 
+    async def is_local_host_in_room_ignoring_users(
+        self, room_id: str, ignore_users: Collection[str]
+    ) -> bool:
+        """Checks if the local server is in the room, ignoring membership of
```
Comment (richvdh) [inline synapse/storage/data_stores/main/roommember.py]: this doesn't read very clearly.

> Check if there are any local users, excluding those in the given list, in the room.

maybe?

File: synapse/storage/data_stores/main/roommember.py [hunk 4]
```diff
@@ -868,6 +868,37 @@ def get_membership_from_event_ids(
...
+                AND NOT (%s)
+                LIMIT 1
+        """ % (
+            clause,
+        )
+
+        def _is_local_host_in_room_ignoring_users_txn(txn):
+            txn.execute(sql, (room_id, Membership.JOIN, *args))
+
+            return bool(txn.fetchone())
```
Comment (richvdh) [inline synapse/storage/data_stores/main/roommember.py]: does `fetchone` do the right thing here when there are 0 rows?

File: synapse/storage/data_stores/main/roommember.py [hunk 5]
```diff
@@ -868,6 +868,37 @@ def get_membership_from_event_ids(
...
+
+        sql = """
+            SELECT 1 FROM local_current_membership
+            WHERE
+                room_id = ? AND membership = ?
+                AND NOT (%s)
+                LIMIT 1
+        """ % (
+            clause,
+        )
```
Comment (richvdh) [inline synapse/storage/data_stores/main/roommember.py]: local_current_membership is populated in a bg update, so this needs to be cunninger.

File: synapse/storage/data_stores/main/roommember.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/storage/data_stores/main/roommember.py]: Yup, you can call `fetchone` repeatedly until it returns `None`

File: synapse/storage/data_stores/main/roommember.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/storage/data_stores/main/roommember.py]: It isn't actually, afaict.

Comment (erikjohnston) [reply synapse/storage/data_stores/main/roommember.py]: https://github.com/matrix-org/synapse/blob/master/synapse/storage/data_stores/main/schema/delta/57/local_current_membership.py

Comment (richvdh) [reply synapse/storage/data_stores/main/roommember.py]: ah fair enough

File: synapse/storage/persist_events.py [hunk 6]
```diff
@@ -736,8 +736,10 @@ def _maybe_start_persisting(self, room_id: str):
                 if event_id == event.event_id:
                     if event.membership == Membership.JOIN:
                         return True
-                else:
-                    events_to_check.append(event_id)
+            else:
```
Comment (richvdh) [inline synapse/storage/persist_events.py]: I think the obscure `else:` here is redundant.

File: synapse/storage/persist_events.py [hunk 7]
```diff
@@ -735,9 +736,35 @@ def _maybe_start_persisting(self, room_id: str):
...
+
+        # Check if any of the changes that we don't have events for are joins.
+        if events_to_check:
+            rows = await self.main_store.get_membership_from_event_ids(events_to_check)
+            is_still_joined = any(row["membership"] == Membership.JOIN for row in rows)
+            if is_still_joined:
+                return True
+
+        # None of the changes to state are joins, so we fall back to checking
+        # the current state of the room to see if any of our users are joined
```
Comment (richvdh) [inline synapse/storage/persist_events.py]: this needs to die
